### PR TITLE
Update to latest rails

### PR DESCRIPTION
**Pull request by zombor**
This PR updates to the latest rails version. Didn't require a lot of changes at all.

I'd like to specify ruby 2.4 in the gemfile, but I don't have it handy locally to test with.

### CODE REVIEW

File: config/application.rb [hunk 1]
```diff
@@ -1,66 +1,18 @@
...
-    # parameters by using an attr_accessible or attr_protected declaration.
-    # config.active_record.whitelist_attributes = true
-
-    # Enable the asset pipeline
-    config.assets.enabled = true
-
-    # Version of your assets, change this if you want to expire all your assets
-    config.assets.version = '1.0'
-
-    if ENV['BASIC_AUTH'] == "true"
```
Comment (christoshrousis) [inline config/application.rb]: Did this move somewhere else and I don't see it? It's required for heroku push to deploy functionality.

File: Gemfile [hunk 2]
```diff
@@ -1,12 +1,12 @@
 source 'https://rubygems.org'
 
-ruby '2.1.4'
+ruby '2.2.6'
```
Comment (christoshrousis) [inline Gemfile]: As per your commentary on the pull request, lets get this up to version 2.4 and merge?

File: config/initializers/session_store.rb [hunk 3]
```diff
@@ -1,8 +1,3 @@
 # Be sure to restart your server when you modify this file.
 
-Elovation::Application.config.session_store :cookie_store, key: '_elovation_session'
-
```
Comment (christoshrousis) [inline config/initializers/session_store.rb]: I've noticed site wide that we're dropping the Elovation namespace, does this have something to do with Rails 5? Drew made it this way, but never specified why do you have any insight on why he may have decided to do it this way?

File: config/application.rb (hunk 1 above)
Comment (zombor) [inline config/application.rb]: Nope! This happened from the rails auto-upgrader. I'll restore it.

**Comment by christoshrousis** [comment]
I can probably address the changes for ruby version and Travis will try revisit asap.

File: config/initializers/session_store.rb (hunk 3 above)
Comment (zombor) [inline config/initializers/session_store.rb]: I'm not sure, honestly. The rails upgrade script changed all these and the app still ran so I kept it.

**Comment by zombor** [comment]
The travis failures are on me, I'll address them and get it passing.

**Comment by zombor** [comment]
Looks like there's lots of changes in tests related to things like rspec rails. I'm working through these...

**Comment by christoshrousis** [comment]
Oh damn. Thanks so much for taking this on though

File: config/environments/production.rb [hunk 4]
```diff
@@ -29,53 +30,62 @@
...
-  # config.autoflush_log = false
-
   # Use default logging formatter so that PID and timestamp are not suppressed.
   config.log_formatter = ::Logger::Formatter.new
+
+  # Use a different logger for distributed setups.
+  # require 'syslog/logger'
+  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
+
+  if ENV["RAILS_LOG_TO_STDOUT"].present?
```
Comment (zombor) [inline config/environments/production.rb]: This replaced the `rails_12factor` gem from before. It's built into rails5.

File: config/puma.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# Puma can serve each request in a thread from an internal thread pool.
```
Comment (zombor) [inline config/puma.rb]: I generally prefer puma, but the default in this app is using unicorn. The rails upgrader created this file, but it should probably be removed?

File: spec/models/sanity_spec.rb [hunk 6]
```diff
@@ -4,13 +4,13 @@
...
     it "ensures deployed migrations have not changed" do
       migrations_digests = {
-        "001_release_001.rb" => "a893705e0162076ba9416f805b13362e",
-        "002_release_002.rb" => "ee6c0915268eca0f8a317c1a7e2048be",
-        "003_release_003.rb" => "ef70f4e0dd6214f426fdd162d56e88c4",
-        "004_release_004.rb" => "152f9acaad9708d938d0d797084c2687",
-        "005_release_005.rb" => "c69855819dbdfa110bc4ab7fe756c565",
-        "006_release_006.rb" => "75c4144504ac63c06a3c089480fbbe14",
-        "007_release_007.rb" => "f30562e00fcdf13f1b9a7bf4430f787e",
+        "001_release_001.rb" => "417a34264a312cff3ee307b900cd87a8",
```
Comment (zombor) [inline spec/models/sanity_spec.rb]: These were needed because of the `[4.2]` addition to the migration files.

File: config/puma.rb (hunk 5 above)
Comment (christoshrousis) [inline config/puma.rb]: Unicorn used to be the default back when this was first created so it makes sense. Lets concentrate on getting the rater sorted and running and then do the puma as a separate pull request - I have a feeling you might need to write some configuration for it? I've only used it once before and the person that handed me the project had already configured it.

**Comment by christoshrousis** [comment]
From reading the codebase I can't tell whether the test is incorrect or whether the code is incorrect and both seem plausible seeing as you're making so many changes. 

I'm leaning more towards the spec if you look at the way it maps the ratings, the values are the same just in the wrong order. I think the results are correct it's just that they're being saved and compared incorrectly?

**Comment by zombor** [comment]
I think the spec seems correct. It's asserting that no matter what order the teams are in, the ratings are the same, which *seems* correct to me.

**Comment by christoshrousis** [comment]
Totally, the need for the spec is correct.

But what if the bump in ruby version and the bump in spec versions is making map behave incorrectly? I would personally re-write the spec to be more exp[licit without the maps and array comparison

I haven't forgotten about your pull request, my work week has been rough - but it should easy up over the next week and I can better address this.

**Comment by christoshrousis** [comment]
Ok mate,

I have a better answer - The spec isn't wrong, you're right. Not only that but the numbers aren't even the same I must be totally blind....

Anyways, I added the following

`game.ratings.sort_by(&:player_id).each{ |r| puts r.inspect }`

on both master and your branch at both ratings points, and it appears that the deviation and mean isn't updating correctly on your branch, on the master branch the deviation and mean are reset - throwing the values out.

The only solution I can think of is to start writing out some specs for expected behaviour of those values and hopefully that bubbles up to find the problem? What do you think?
